### ISSUE TITLE
Calling base constructors from derived classes in module results in a stack overflow

### ISSUE THREAD

**Issue by DezZolation** [yanghuan/CSharp.lua#205]
Consider the following class inheritence structure.

```cs
using System;

namespace FooSpace {
  public class A
  {
      public A(string foo){}
      public A(int foo){}
  }

  public class B : A
  {
      public B(string foo) : base(foo){}
      public B(int foo) : base(foo){}
  }

  public class C : B
  {
      public C(string foo) : base(foo){}
      public C(int foo) : base(foo){}
  }
}
```

This will parse to a Lua file where the constructors of `B` will reference `A` by its full class while the constructors of `C` will reference `B` using ``System.base(this)`` 

```Lua
namespace.class("B", function (namespace)
    local __ctor1__, __ctor2__
    __ctor1__ = function (this, foo)
      FooSpace.A.__ctor__[1](this, foo)
    end
    __ctor2__ = function (this, foo)
      FooSpace.A.__ctor__[2](this, foo)
    end
    return {
      __inherits__ = function (out)
        return {
          out.FooSpace.A
        }
      end,
      __ctor__ = {
        __ctor1__,
        __ctor2__
      }
    }
  end)

  namespace.class("C", function (namespace)
    local __ctor1__, __ctor2__
    __ctor1__ = function (this, foo)
      System.base(this).__ctor__[1](this, foo)
    end
    __ctor2__ = function (this, foo)
      System.base(this).__ctor__[2](this, foo)
    end
    return {
      __inherits__ = function (out)
        return {
          out.FooSpace.B
        }
      end,
      __ctor__ = {
        __ctor1__,
        __ctor2__
      }
    }
  end)
```
I deduce that the CSharp.lua checks if a class is ever inherited from in the project, if that is the case then it changes the base constructor to a full class reference. (``FooSpace.A.``)

However, when I were to put class `C` in a seperate module. Class `B` suddenly receives the ``System.base(this)`` base constructor.

```Lua
namespace.class("B", function (namespace)
    local __ctor1__, __ctor2__
    __ctor1__ = function (this, foo)
      System.base(this).__ctor__[1](this, foo)
    end
    __ctor2__ = function (this, foo)
      System.base(this).__ctor__[2](this, foo)
    end
    return {
      __inherits__ = function (out)
        return {
          out.FooSpace.A
        }
      end,
      __ctor__ = {
        __ctor1__,
        __ctor2__
      }
    }
  end)
```
Also relevant, ``System.base()``:
```Lua
function System.base(this)
  return getmetatable(getmetatable(this))
end
```
Because class `C` still calls `B`'s constructors. ``this`` is passed as an argument to ``System.base()`` in `B`, which returns the same constructor again. This results in an infinite loop. This is the pseudo callstack that I deduce:
```
C.__ctor1__ () -> 
System.base(this)__ctor[1]__() -> 
B.__ctor1__ ->
System.base(this)__ctor[1]__() -> 
B.__ctor1__ ->
System.base(this)__ctor[1]__() -> 
... etc.
```
Possible solution: Don't use ``System.base()`` to find the base constructor but always use the full class reference.

**Comment by yanghuan** [comment]
thands report, i fix it by d0810fe510f3f49af9d554fca17a9d9e22419399
you should use -module arguent to compile the code whitch need be referenced

**Comment by DezZolation** [comment]
I have tested it in our project and it works. Our gratitude is eternal!
